### DISCUSSION TITLE
Builds successfully for Android, but fails for iOS: firebase_performance 0.10.0-8 does not play nice with firebase_core

### DISCUSSION THREAD

**Discussion by oneForCode** [codemagic-ci-cd/codemagic-docs#2798] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

Since the FlutterFlow dependency update I exported the code and tired to build for Android and iOS, same as I always did.

### Expected results

It used to build fine for both iOS and Android. But since dependency update only Android builds successfully;

### Actual results

    [!] CocoaPods could not find compatible versions for pod "FirebaseCore":
      In Podfile:
        firebase_core (from `.symlinks/plugins/firebase_core/ios`) was resolved to 3.6.0, which depends on
          Firebase/CoreOnly (= 11.2.0) was resolved to 11.2.0, which depends on
            FirebaseCore (= 11.2.0)

        firebase_performance (from `.symlinks/plugins/firebase_performance/ios`) was resolved to 0.10.0-8, which depends on
          Firebase/Performance (= 11.2.0) was resolved to 11.2.0, which depends on
            FirebasePerformance (~> 11.2.0) was resolved to 11.2.0, which depends on
              FirebaseCore (~> 11.0)

### Build id (optional)

671b37e6dbb3fcd02b0ec79c

**Comment by dam-ease** [comment]
Hello @oneForCode,
From the error messages in your logs, there are version mismatches/conflicts in your `Podfile` and `Podfile.lock` file for the Firebase/CoreOnly, Firebase/Performance and GoogleMLKit/BarcodeScanning plugins.
Please add this below script in the Pre-build section of your Workflow UI, it should remove and automatically generate a new podfile for you.
```
​cd ios
rm Podfile.lock
rm -rf Pods
pod install
``` 
And re-run the build, let us know how it goes.

**Reply by oneForCode** [reply]
Thank you so much for your reply @dam-ease!

I tried it. Build ID: 671f82a88b5d0298fce32015

![Screenshot from 2024-10-28 06-24-46](https://github.com/user-attachments/assets/4d997863-909b-4052-8fd9-4e7606ff11d1)
![Screenshot from 2024-10-28 06-35-49](https://github.com/user-attachments/assets/4c0f96b9-50ea-4a71-9234-bd681a9cce3f)


which makes sense because my ios folder does not have Podfile.lock

![Screenshot from 2024-10-28 06-33-15](https://github.com/user-attachments/assets/9f746048-1f5f-4afd-8007-f48ace3ba320)

Any other suggestions? Thank you again!

**Reply by dam-ease** [reply]
Hey @oneForCode.
Please run `flutter create .` locally to generate the missing file, push the updated code and try starting a new build. This should work!

**Reply by oneForCode** [reply]
Thanks @dam-ease!

Same result.

Build ID 6721079b173a4dfbdc07bc3c   (without Pre-build script)

> 
    [!] CocoaPods could not find compatible versions for pod "FirebaseCore":
      In Podfile:
        firebase_core (from `.symlinks/plugins/firebase_core/ios`) was resolved to 3.6.0, which depends on
          Firebase/CoreOnly (= 11.2.0) was resolved to 11.2.0, which depends on
            FirebaseCore (= 11.2.0)

        firebase_performance (from `.symlinks/plugins/firebase_performance/ios`) was resolved to 0.10.0-8, which depends on
          Firebase/Performance (= 11.2.0) was resolved to 11.2.0, which depends on
            FirebasePerformance (~> 11.2.0) was resolved to 11.2.0, which depends on
              FirebaseCore (~> 11.0)

OR

Build ID 6721099ce4926ae0f9299606 (with your suggested Pre-build script)

![Screenshot from 2024-10-29 10-16-26](https://github.com/user-attachments/assets/840a6fda-5b5a-4143-88eb-98dd0ba7ca21)

Thanks for your help!

**Reply by dam-ease** [reply]
Hello @oneForCode,
Do you have the `podfile.lock` in your directory now?
You can add below script to your pre-build script instead:
```
​find . -name "Podfile" -execdir pod update \;
```
Also there is a solution to this in this SO thread - https://stackoverflow.com/a/65808257/11718149, this will help resolve this as well.

**Comment by oneForCode** [comment]
Hello @dam-ease,

Thanks for your response.

I do have `podfile.lock` now, because I did  `flutter create .` locally, and then checked it in.

I did run `​find . -name "Podfile" -execdir pod install \;` as pre-build script as you suggested 

![Screenshot from 2024-10-30 09-54-54](https://github.com/user-attachments/assets/11485233-2722-41f2-83e9-bc97367322e2)

and get the same error:

> 
    [!] CocoaPods could not find compatible versions for pod "FirebaseCore":
      In Podfile:
        firebase_core (from `.symlinks/plugins/firebase_core/ios`) was resolved to 3.6.0, which depends on
          Firebase/CoreOnly (= 11.2.0) was resolved to 11.2.0, which depends on
            FirebaseCore (= 11.2.0)

        firebase_performance (from `.symlinks/plugins/firebase_performance/ios`) was resolved to 0.10.0-8, which depends on
          Firebase/Performance (= 11.2.0) was resolved to 11.2.0, which depends on
            FirebasePerformance (~> 11.2.0) was resolved to 11.2.0, which depends on
              FirebaseCore (~> 11.0)

What I don't understand, given that: 

> FirebasePerformance (\~> 11.2.0) was resolved to 11.2.0, which depends on
          FirebaseCore (~> 11.0)

the problem seems to be that correct version of `FirebasePerformance` depends on incorrect version of `FirebaseCore`.

How does pre-building it helps, since Codemagic is already building it? 

Isn't the problem in `firebase_performance` library itself?

Thanks!

**Reply by dam-ease** [reply]
Hey @oneForCode,
Yes, you are right with problem!
If you run `pod install` locally on your machine, you should be able to reproduce this issue as well.

The idea of running `find . -name "Podfile" -execdir pod update \;` here is to update the versions of these packages to the latest compatible versions such that the compatibility issue is resolved. 
Please note that this can also be done manually by adjusting the versions of the packages in the `podfile.lock` file.

On my end, adding the latest versions of `firebase_performance` and `firebase_core` packages does't give any issues. Are you making use of the latest versions as well? Please can you share your `pubspec.yaml` file as well as the generated `podfile.lock` here, so I can try reproducing on my end as well?

**Reply by oneForCode** [reply]
Thanks you @dam-ease!
In the `pubspec.yaml`

> version: 1.0.0+1
> 
> environment:
>   sdk: ">=3.0.0 <4.0.0"
> 
> dependencies:
>   flutter:
>     sdk: flutter
>   flutter_localizations:
>     sdk: flutter
>   flutter_web_plugins:
>     sdk: flutter
>   ai_barcode_scanner: ^5.2.2
>   audio_session: 0.1.21
>   auto_size_text: 3.0.0
>   cached_network_image: 3.4.1
>   cached_network_image_platform_interface: 4.1.1
>   cached_network_image_web: 1.3.1
>   cloud_firestore: 5.4.4
>   cloud_firestore_platform_interface: 6.4.3
>   cloud_firestore_web: 4.3.2
>   cloud_functions: 5.1.3
>   cloud_functions_platform_interface: 5.5.37
>   cloud_functions_web: 4.10.2
>   collection: 1.18.0
>   csslib: 0.17.3
>   easy_debounce: 2.0.1
>   file_picker: 8.1.2
>   firebase_auth: 5.3.1
>   firebase_auth_platform_interface: 7.4.7
>   firebase_auth_web: 5.13.2
>   firebase_core: 3.6.0
>   firebase_core_platform_interface: 5.3.0
>   firebase_core_web: 2.18.1
>   firebase_crashlytics: 4.1.3
>   firebase_crashlytics_platform_interface: 3.6.44
>   firebase_performance: 0.10.0+8
>   firebase_performance_platform_interface: 0.1.4+44
>   firebase_performance_web: 0.1.7+2
>   firebase_remote_config: 5.1.3
>   firebase_remote_config_platform_interface: 1.4.44
>   firebase_remote_config_web: 1.7.2
>   firebase_storage: 12.3.2
>   firebase_storage_platform_interface: 5.1.31
>   firebase_storage_web: 3.10.2
>   flutter_animate: 4.5.0
>   flutter_barcode_scanner: 
>     git:
>       url: https://github.com/FlutterFlow/flutter_barcode_scanner
>       ref: v2.0.0
>   flutter_cache_manager: 3.4.1
>   flutter_keyboard_visibility: 6.0.0
>   flutter_keyboard_visibility_linux: 1.0.0
>   flutter_keyboard_visibility_macos: 1.0.0
>   flutter_keyboard_visibility_platform_interface: 2.0.0
>   flutter_keyboard_visibility_web: 2.0.0
>   flutter_keyboard_visibility_windows: 1.0.0
>   flutter_plugin_android_lifecycle: 2.0.20
>   font_awesome_flutter: 10.7.0
>   from_css_color: 2.0.0
>   go_router: 12.1.3
>   google_fonts: 6.1.0
>   google_sign_in: 6.2.1
>   google_sign_in_android: 6.1.30
>   google_sign_in_ios: 5.7.7
>   google_sign_in_platform_interface: 2.4.5
>   google_sign_in_web: 0.12.4+2
>   html: 0.15.4
>   image_picker: 1.1.2
>   image_picker_android: 0.8.12+13
>   image_picker_for_web: 3.0.5
>   image_picker_ios: 0.8.12
>   image_picker_linux: 0.2.1+1
>   image_picker_macos: 0.2.1+1
>   image_picker_platform_interface: 2.10.0
>   image_picker_windows: 0.2.1+1
>   internet_connection_checker_plus: ^2.5.2
>   intl: 0.19.0
>   json_path: 0.7.2
>   just_audio: 0.9.40
>   just_audio_platform_interface: 4.3.0
>   just_audio_web: 0.4.12
>   mime_type: 1.0.0
>   package_info_plus: ^8.0.2
>   page_transition: 2.1.0
>   path_provider: 2.1.4
>   path_provider_android: 2.2.10
>   path_provider_foundation: 2.4.0
>   path_provider_linux: 2.2.1
>   path_provider_platform_interface: 2.1.2
>   path_provider_windows: 2.2.1
>   plugin_platform_interface: 2.1.8
>   pointer_interceptor: 0.10.1+2
>   pointer_interceptor_ios: 0.10.1
>   pointer_interceptor_platform_interface: 0.10.0+1
>   pointer_interceptor_web: 0.10.2+1
>   provider: 6.1.2
>   rxdart: 0.27.7
>   share_plus: 10.0.2
>   share_plus_platform_interface: 5.0.0
>   shared_preferences: 2.3.2
>   shared_preferences_android: 2.3.2
>   shared_preferences_foundation: 2.5.2
>   shared_preferences_linux: 2.4.1
>   shared_preferences_platform_interface: 2.4.1
>   shared_preferences_web: 2.4.2
>   shared_preferences_windows: 2.4.1
>   sign_in_with_apple: 6.1.2
>   sign_in_with_apple_platform_interface: 1.1.0
>   sign_in_with_apple_web: 2.1.0
>   smooth_page_indicator: 1.1.0
>   sqflite: 2.3.3+1
>   sqflite_common: 2.5.4+3
>   stepper_counter_swipe: ^1.0.2
>   stream_transform: 2.1.0
>   timeago: 3.6.1
>   url_launcher: 6.3.0
>   url_launcher_android: 6.3.10
>   url_launcher_ios: 6.3.1
>   url_launcher_linux: 3.2.0
>   url_launcher_macos: 3.2.1
>   url_launcher_platform_interface: 2.3.2
>   url_launcher_web: 2.3.3
>   url_launcher_windows: 3.1.2
>   uuid: ^4.0.0
>   video_player: 2.9.2
>   video_player_android: 2.7.13
>   video_player_avfoundation: 2.6.2
>   video_player_platform_interface: 6.2.3
>   video_player_web: 2.3.2
>   visibility_detector: ^0.4.0+2
>   wakelock_plus: ^1.2.8
>   webview_flutter: 4.9.0
>   webview_flutter_android: 3.16.7
>   webview_flutter_platform_interface: 2.10.0
>   webview_flutter_wkwebview: 3.15.0
>   webviewx_plus: 0.5.0
> 
> 
> dependency_overrides:
>   http: 1.2.2
>   pointer_interceptor: 0.10.1+2
>   pointer_interceptor_ios: 0.10.1
>   pointer_interceptor_platform_interface: 0.10.0+1
>   pointer_interceptor_web: 0.10.2+1
>   uuid: ^4.0.0
>   win32: 5.5.1
> 
> dev_dependencies:
>   flutter_launcher_icons: 0.13.1
>   flutter_lints: 4.0.0
>   image: 4.2.0
>   lints: 4.0.0
>   flutter_test:
>     sdk: flutter

**Comment by oneForCode** [comment]
@dam-ease, Thank you so much for you help.

The problem is solved. 
The problem was that in `Podfile` FlutterFlow was using the `firestore-ios-sdk-frameworks` binary distribution while also having the regular Firebase pods installed. This caused conflicts.

I modified the `Podfile` as follows:

> \# Add this line at the top
> $FirebaseSDKVersion = '11.2.0'
> 
> \# Add these lines before flutter_install_all_ios_pods
> pod 'Firebase', :modular_headers => true
> pod 'FirebaseCore', :modular_headers => true
> pod 'GoogleUtilities', :modular_headers => true
> 
> \# changing:
> pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '11.2.0'
> to
> pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :branch => "11.2.0"

Thanks again for your kind help! All the best to you!
